crates/evm-tracing-client/src/listeners/call_list.rs

```diff
@@ -0,0 +1,1098 @@
+
+use evm_tracing_events::{
+    runtime::{Capture, ExitError, ExitReason, ExitSucceed},
+    Event, EvmEvent, GasometerEvent, Listener as ListenerT, RuntimeEvent, StepEventFilter,
+};
+use sp_core::{sp_std::collections::btree_map::BTreeMap, H160, U256};
+
+use crate::types::{
+    blockscout::{BlockscoutCall as Call, BlockscoutCallInner as CallInner},
+    CallResult, CallType, ContextType, CreateResult,
+};
+
+/// Enum of the different "modes" of tracer for multiple runtime versions and
+/// the kind of EVM events that are emitted.
+enum TracingVersion {
+    /// The first event of the transaction is `EvmEvent::TransactX`. It goes along other events
+    /// such as `EvmEvent::Exit`. All contexts should have clear start/end boundaries.
+    EarlyTransact,
+    /// Older version in which the events above didn't existed.
+    /// It means that we cannot rely on those events to perform any task, and must rely only
+    /// on other events.
+    Legacy,
+}
+
+/// Listener.
+pub struct Listener {
+    /// Version of the tracing.
+    /// Defaults to legacy, and switch to a more modern version if recently added events are
+    /// received.
+    version: TracingVersion,
+    /// Transaction cost that must be added to the first context cost.
+    transaction_cost: u64,
+    /// Final logs.
+    pub entries: Vec<BTreeMap<u32, Call>>,
+    /// Next index to use.
+    entries_next_index: u32,
+    /// Stack of contexts with data to keep between events.
+    context_stack: Vec<Context>,
+    /// Type of the next call.
+    /// By default is None and corresponds to the root call, which
+    /// can be determined using the `is_static` field of the `Call` event.
+    /// Then by looking at call traps events we can set this value to the correct
+    /// call type, to be used when the following `Call` event is received.
+    call_type: Option<CallType>,
+    /// When `EvmEvent::TransactX` is received it creates its own context. However it will usually
+    /// be followed by an `EvmEvent::Call/Create` that will also create a context, which must be
+    /// prevented. It must however not be skipped if `EvmEvent::TransactX` was not received
+    /// (in legacy mode).
+    skip_next_context: bool,
+    /// To handle `EvmEvent::Exit` no emitted by previous runtimes versions,
+    /// entries are not inserted directly in `self.entries`.
+    /// `pending_entries`: Vec<(u32, Call)>,
+    /// See `RuntimeEvent::StepResult` event explanatioins.
+    step_result_entry: Option<(u32, Call)>,
+    /// When tracing a block `Event::CallListNew` is emitted before each Ethereum transaction is
+    /// processed. Since we use that event to **finish** the transaction, we must ignore the first
+    /// one.
+    call_list_first_transaction: bool,
+    /// True if only the `GasometerEvent::RecordTransaction` event has been received.
+    /// Allow to correctly handle transactions that cannot pay for the tx data in Legacy mode.
+    record_transaction_event_only: bool,
+}
```

Comment:
I didn't understand why we don't need [`with_log`](https://github.com/moonbeam-foundation/moonbeam/blob/4a7944905aa8b6d5e897e2c1f209f4468747b379/client/evm-tracing/src/listeners/call_list.rs#L84). Reviewed the discussion history. Did we decide not to support `EvmEvent::Log`?

Reply:
Please, check the `evm` version we use. It doesn't contain `Log` logic. 

Additionally, we can't just bump it as it should be compatible with our frontier fork.

Reply:
https://github.com/rust-ethereum/evm/blob/b7b82c7e1fc57b7449d6dfa6826600de37cc1e65/src/tracing.rs#L14